geomesa-native-api/src/main/scala/org/locationtech/geomesa/api/AccumuloGeoMesaIndex.scala

```diff
@@ -0,0 +1,103 @@
+package org.locationtech.geomesa.api
+
+import java.lang.Iterable
+import java.util.Date
+
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.vividsolutions.jts.geom.Geometry
+import org.geotools.data.simple.SimpleFeatureWriter
+import org.geotools.data.{DataStoreFinder, Transaction}
+import org.geotools.factory.CommonFactoryFinder
+import org.geotools.filter.identity.FeatureIdImpl
+import org.geotools.filter.text.cql2.CQL
+import org.locationtech.geomesa.accumulo.data.tables.{RecordTable, Z3Table}
+import org.locationtech.geomesa.accumulo.data.{AccumuloDataStore, AccumuloDataStoreParams}
+import org.locationtech.geomesa.accumulo.util.Z3UuidGenerator
+import org.locationtech.geomesa.utils.geotools.{Conversions, SftBuilder}
+
+class AccumuloGeoMesaIndex[T](ds: AccumuloDataStore,
+                              fname: String,
+                              serde: ValueSerializer[T]) extends GeoMesaIndex[T] {
+  private val ff = CommonFactoryFinder.getFilterFactory2
+
+  if(!ds.getTypeNames.contains(fname)) {
+    val sft =
+      new SftBuilder()
+        .date("dtg", default = true, index = true)
+        .bytes("payload")
+        .geometry("geom", default = true)
+        .withIndexes(List(Z3Table.suffix, RecordTable.suffix))
+        .userData("geomesa.mixed.geometries","true")
+        .build(fname)
+    ds.createSchema(sft)
+  }
+
+  val fs = ds.getFeatureSource(fname)
+
+  val writers =
+    CacheBuilder.newBuilder().build(
+      new CacheLoader[String, SimpleFeatureWriter] {
+        override def load(k: String): SimpleFeatureWriter = {
+          ds.getFeatureWriterAppend(k, Transaction.AUTO_COMMIT).asInstanceOf[SimpleFeatureWriter]
+        }
+      })
+
+  override def query(query: GeoMesaQuery): Iterable[T] = {
+    import Conversions._
+
+    import scala.collection.JavaConverters._
+
+    fs.getFeatures(query.getFilter)
+      .features()
+      .map { f => serde.fromBytes(f.getAttribute(1).asInstanceOf[Array[Byte]]) }
+      .toIterable.asJava
+  }
+
+  override def insert(value: T, geom: Geometry, dtg: Date): String = {
+    val bytes = serde.toBytes(value)
+    val fw = writers.get(fname)
+    val sf = fw.next()
+    val id = Z3UuidGenerator.createUuid(geom, dtg.getTime).toString
```

Comment:
Do we intend to support user-provided IDs?  Clearly, the serialized native object will contain its own ID, but aren't there use-cases in which the consumer wants to query based on the native IDs rather than our synthetic ones?